src/lib/OpenEXR/ImfHeader.cpp

```diff
@@ -60,8 +62,112 @@ using IMATH_NAMESPACE::Box2i;
 using IMATH_NAMESPACE::V2i;
 using IMATH_NAMESPACE::V2f;
 
+namespace
+{
+
+static int   s_DefaultZipCompressionLevel = Z_DEFAULT_COMPRESSION;
+static float s_DefaultDwaCompressionLevel = 45.f;
+
+struct CompressionRecord
+{
+    CompressionRecord()
+        : zip_level (s_DefaultZipCompressionLevel),
+          dwa_level (s_DefaultDwaCompressionLevel)
+    {}
+    int zip_level;
+    float dwa_level;
+};
+// NB: This is extra complicated than one would normally write to
+// handle scenario that seems to happen on MacOS with a static
+// library, where the header is being destroyed after the shutdown of
+// this translation unit happens, causing use after destroy.
+//
+// but if we just use the once_flag / call_once mechanism, windows
+// then starts crashing on exit in a different way.
 
-namespace {
+struct CompressionStash;
+// assignments to here happen either in static singleton ctor or
+// shutdown should be thread safe
+static CompressionStash *s_stash = nullptr;
+
+struct CompressionStash
+{
+    CompressionStash()
+    {
+        s_stash = this;
+    }
+    ~CompressionStash()
+    {
+        s_stash = nullptr;
```

Comment:
I think this needs to be protected by the mutex, in order to stall destruction of the map in the case that the mutex has already been acquired, for example by an ImfHeader destructor trying to remove itself from the stash.

Reply:
well, it's at app static object cleanup time that's the issue, and only when compiling statically (non-deterministic translation unit shutdown ordering) - the first version of this I was getting crashes trying to lock a destroyed mutex, so not sure the extra checks are actually going to help. However, I forgot to check whether threading was enabled or not in the library, so will add those ifdefs and a safety lock / clear just to be a bit safer.